### PR TITLE
Access control filter for the gateway

### LINKED ISSUE #2954: The Gateway should only forward the "/api" and Swagger URLs
Currently the gateway forwards applications automatically, which is the default with Zuul.

However, for security reasons, we shouldn't forward everything, like all the "/info" URLs.

So the gateway should be reconfigured to only forward URLs to:

- "/api": where all the application's APIs are
- "/v2/api-docs": where Swagger is -> this should only work in "dev" mode, not "prod" mode (still for security reasons)

### PR DESCRIPTION
**Pull request by PierreBesson**
Fix #2954 

It can be configured by changing properties under jhipster.gateway.authorized-microservices-endpoints.
By default, all microservices endpoints are available (spring cloud default configuration) but a sample recommended config is provided for dev and prod.

### CODE REVIEW

File: generators/server/templates/src/main/java/package/gateway/accesscontrol/_AccessControlFilter.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import org.springframework.http.HttpStatus;
+
+import javax.inject.Inject;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Zuul filter for restricting access to backend micro-services endpoints
+ */
+public class AccessControlFilter extends ZuulFilter{
```
Comment (jdubois) [inline generators/server/templates/src/main/java/package/gateway/accesscontrol/_AccessControlFilter.java]: Formatting: you miss a space at `ZuulFilter {`

File: generators/server/templates/src/main/java/package/gateway/accesscontrol/_AccessControlFilter.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        }
+        return true;
+    }
+
+    private boolean isAuthorizedRequest(String serviceUrl, String serviceName, String requestUri) {
+        Map<String, List<String>> authorizedMicroservicesEndpoints = jHipsterProperties.getGateway().getAuthorizedMicroservicesEndpoints();
+
+        // If the authorized endpoints list was left empty for this route, all access are allowed
+        if(authorizedMicroservicesEndpoints.get(serviceName) == null){
+            log.debug("Access Control: allowing access for: " + requestUri + ", as no access control policy has been set up for service: " + serviceName);
```
Comment (jdubois) [inline generators/server/templates/src/main/java/package/gateway/accesscontrol/_AccessControlFilter.java]: For the log.debug you need to use the `log.debug("my text {}", value);` pattern as it's more efficient

File: generators/server/templates/src/main/java/package/gateway/accesscontrol/_AccessControlFilter.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            return true;
+        }
+        else {
+            List<String> authorizedEndpoints = authorizedMicroservicesEndpoints.get(serviceName);
+
+            // Go over the authorized endpoints to control that the request URI matches it
+            for (String endpoint : authorizedEndpoints) {
+                String gatewayEndpoint = serviceUrl.substring(0, serviceUrl.length() - 3) + endpoint;
+                if (requestUri.startsWith(gatewayEndpoint)) {
+                    log.debug("Access Control: allowing access for: " + requestUri + ", as it matches the following authorized microservice endpoint: " + gatewayEndpoint);
```
Comment (jdubois) [inline generators/server/templates/src/main/java/package/gateway/accesscontrol/_AccessControlFilter.java]: same here for logging

File: generators/server/templates/src/main/java/package/gateway/accesscontrol/_AccessControlFilter.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.netflix.zuul.filters.ProxyRouteLocator;
+import org.springframework.http.HttpStatus;
+
+import javax.inject.Inject;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Zuul filter for restricting access to backend micro-services endpoints
```
Comment (jdubois) [inline generators/server/templates/src/main/java/package/gateway/accesscontrol/_AccessControlFilter.java]: you need a "." at the end to have correct Javadoc

**Comment by deepu105** [comment]
LGTM +1

**Comment by jdubois** [comment]
OK let's do this :-)

**Pull request merged**